pgjdbc/src/main/java/org/postgresql/util/LazyCleaner.java

```diff
@@ -0,0 +1,220 @@
+
+package org.postgresql.util;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.lang.ref.PhantomReference;
+import java.lang.ref.ReferenceQueue;
+import java.time.Duration;
+import java.util.concurrent.ThreadFactory;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * LazyCleaner is a utility class that allows to register objects for deferred cleanup.
+ * <p>Note: this is a driver-internal class</p>
+ */
+public class LazyCleaner {
+  private static final Logger LOGGER = Logger.getLogger(LazyCleaner.class.getName());
+  private static final LazyCleaner instance =
+      new LazyCleaner(
+          Duration.ofMillis(Long.getLong("pgjdbc.config.cleanup.thread.ttl", 30000)),
+          "pgjdbc cleanup thread"
+      );
+
+  public interface Cleanable<T extends Throwable> {
+    void clean() throws T;
+  }
+
+  public interface CleaningAction<T extends Throwable> {
+    void onClean(boolean leak) throws T;
+  }
+
+  private final ReferenceQueue<Object> queue = new ReferenceQueue<>();
+  private final long threadTtl;
+  private final ThreadFactory threadFactory;
+  private boolean threadRunning = false;
+  private int watchedCount = 0;
+  private @Nullable Node<?> first;
+
+  /**
+   * Returns a default cleaner instance.
+   * <p>Note: this is driver-internal API.</p>
+   * @return the instance of LazyCleaner
+   */
+  public static LazyCleaner getInstance() {
+    return instance;
+  }
+
+  public LazyCleaner(Duration threadTtl, final String threadName) {
+    this(threadTtl, runnable -> {
+      Thread thread = new Thread(runnable, threadName);
+      thread.setDaemon(true);
+      return thread;
+    });
+  }
+
+  private LazyCleaner(Duration threadTtl, ThreadFactory threadFactory) {
+    this.threadTtl = threadTtl.toMillis();
+    this.threadFactory = threadFactory;
+  }
+
+  public <T extends Throwable> Cleanable<T> register(Object obj, CleaningAction<T> action) {
+    return add(new Node<T>(obj, action));
+  }
+
+  public synchronized int getWatchedCount() {
+    return watchedCount;
+  }
+
+  public synchronized boolean isThreadRunning() {
+    return threadRunning;
+  }
+
+  private synchronized boolean checkEmpty() {
+    if (first == null) {
+      threadRunning = false;
+      return true;
+    }
+    return false;
+  }
+
+  private synchronized <T extends Throwable> Node<T> add(Node<T> node) {
+    if (first != null) {
+      node.next = first;
+      first.prev = node;
+    }
+    first = node;
+    watchedCount++;
+
+    if (!threadRunning) {
+      threadRunning = startThread();
+    }
+    return node;
+  }
+
+  private boolean startThread() {
+    Thread thread = threadFactory.newThread(new Runnable() {
+      public void run() {
+        while (true) {
+          try {
+            // Clear setContextClassLoader to avoid leaking the classloader
+            Thread.currentThread().setContextClassLoader(null);
+            Thread.currentThread().setUncaughtExceptionHandler(null);
```

Comment:
@davecramer , this was leaking a classloader.
I'm afraid the thread we create will inherit thread locals, so we still can leak memory. I am not sure we want to go for "reflection-based approach to clean thread locals"

It might be the better way out would be using `Cleaner` on Java 9+ and finalize on Java 8.

Reply:
> It might be the better way out would be using `Cleaner` on Java 9+ and finalize on Java 8.

That's going to complicate things... but sure. Java 8 should be deprecated soon(ish)